src/Tutor.Core/LearnerModel/DomainOverlay/KnowledgeComponentMasteries/SessionTracker.cs

```diff
@@ -0,0 +1,84 @@
+﻿using FluentResults;
+using System;
+using Tutor.Core.BuildingBlocks.EventSourcing;
+using Tutor.Core.LearnerModel.DomainOverlay.KnowledgeComponentMasteries.Events;
+using Tutor.Core.LearnerModel.DomainOverlay.KnowledgeComponentMasteries.Events.SessionLifecycleEvents;
+
+namespace Tutor.Core.LearnerModel.DomainOverlay.KnowledgeComponentMasteries
+{
+    public class SessionTracker : EventSourcedEntity
+    {
+        public TimeSpan DurationOfPreviousSessions { get; private set; } = new TimeSpan(0);
+        public TimeSpan DurationOfCurrentSession { get => HasActiveSession ? LastActivity.Value - StartOfCurrentSession.Value : new TimeSpan(0); }
+        public bool HasActiveSession { get => StartOfCurrentSession.HasValue; }
+        public DateTime? StartOfCurrentSession { get; private set; }
+        public DateTime? LastActivity { get; private set; }
+
+        public void Launch()
+        {
+            if (HasActiveSession) Causes(new SessionAbandoned());
+
+            Causes(new SessionLaunched());
+        }
+
+        public Result Terminate()
+        {
+            if (!HasActiveSession) return Result.Fail("No active session to terminate.");
+
+            Causes(new SessionTerminated());
+            return Result.Ok();
+        }
+
+        public Result Abandon()
+        {
+            if (!HasActiveSession) return Result.Fail("No active session to abandon.");
+
+            Causes(new SessionAbandoned());
+            return Result.Ok();
+        }
+
+        public override void Apply(DomainEvent @event)
+        {
+            When((dynamic)@event);
+        }
+
+        private void When(SessionLaunched @event)
+        {
+            StartOfCurrentSession = @event.TimeStamp;
+            LastActivity = @event.TimeStamp;
+
+            @event.DurationOfPreviousSessions = DurationOfPreviousSessions;
+            @event.DurationOfCurrentSession = DurationOfCurrentSession;
+        }
+
+        private void When(SessionTerminated @event)
+        {
+            LastActivity = @event.TimeStamp;
+
+            @event.DurationOfPreviousSessions = DurationOfPreviousSessions;
+            @event.DurationOfCurrentSession = DurationOfCurrentSession;
+
+            DurationOfPreviousSessions += DurationOfCurrentSession;
+            StartOfCurrentSession = null;
+            LastActivity = null;
+        }
+
+        private void When(SessionAbandoned @event)
+        {
+            @event.DurationOfPreviousSessions = DurationOfPreviousSessions;
+            @event.DurationOfCurrentSession = DurationOfCurrentSession;
+
+            DurationOfPreviousSessions += DurationOfCurrentSession;
+            StartOfCurrentSession = null;
+            LastActivity = null;
+        }
+
+        private void When(KnowledgeComponentEvent @event)
+        {
+            LastActivity = @event.TimeStamp;
+
+            @event.DurationOfPreviousSessions = DurationOfPreviousSessions;
```

Comment:
It would be helpful if I can see a sequence diagram (simplest possible sketch is enough) to demonstrate how the service invokes these events for the more complex design of the event sourced aggregate.

Reply:
I'm not really sure what to include in the sequence diagram? As far as the service is concerned everything stays the same, it just invokes public command methods on the aggregate root.
Or do you mean a diagram of what a command method of the aggregate root would look like in cases where events are caused by an entity or need to be applied to an entity?

Reply:
Yes, I was aiming for the scenario where the root and the entity are interacting through events.

Reply:
Basically, the only two possible interactions are:

- The aggregate root can delegate some part of a command method to an entity by calling one of the entity's command methods. If the entity method needs to cause an event, that's achieved by calling the Causes method on the root, so events caused by entities are treated the same as events caused by the root.
- When updating the aggregate state based on an event through the root's Apply method, an entity might need to be updated. That's achieved by passing the event to the entity's Apply method. 

Below are sequence diagrams illustrating a simple case of those two interactions: 

![1](https://user-images.githubusercontent.com/57844740/191348939-1d3e035b-62de-4344-bd74-d3a9ba4d595a.png)
![2](https://user-images.githubusercontent.com/57844740/191348416-8f68df41-2721-4e0c-9b0c-4b8aff607fbb.png)